pkg/sql/catalog/tabledesc/logical_replication_helpers.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tabledesc
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/errors"
+)
+
+const cannotLDRMsg = "cannot create logical replication stream"
+
+// CheckLogicalReplicationCompatibility verifies that the destination table
+// descriptor is a valid target for logical replication from the source table.
+func CheckLogicalReplicationCompatibility(src, dst *descpb.TableDescriptor) error {
+	if err := checkSrcDstColsMatch(src, dst); err != nil {
+		return err
+	}
+
+	if err := checkColumnFamilies(dst); err != nil {
+		return err
+	}
+
+	if err := checkCompositeTypesInPrimaryKey(dst); err != nil {
+		return err
+	}
+
+	if err := checkVirtualColumns(dst); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// We disallow expression evaluation (e.g. virtual columns that are stored in an
+// index) because the LDR KV write path does not understand how to evaluate
+// expressions. The writer expects to receive the full set of columns, even the
+// computed ones, along with a list of columns that we've already determined
+// should be updated.
+func checkVirtualColumns(dst *descpb.TableDescriptor) error {
+	// Disallow partial indexes.
+	for _, idx := range dst.Indexes {
+		if idx.IsPartial() {
+			return errors.WithMessage(pgerror.Newf(
+				pgcode.InvalidTableDefinition,
+				"table %s has a partial index %s", dst.Name, idx.Name,
+			), cannotLDRMsg)
+		}
+	}
+
+	// Disallow virtual columns if they are a key of an index.
+	// NB: it is impossible for a virtual column to be stored in an index.
+	for _, col := range dst.Columns {
```

Comment:
nit, it may not matter, but @stevendanna had a similar implementation of this check that avoided nesting a for loop by building a `ColumnIDToOrdinalMap`: https://github.com/cockroachdb/cockroach/blob/82dc197abc1a15e001f6666c82e0a7c709cbec1c/pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go#L288

I'm not sure why why he used only `PublicColumns()` though.

Reply:
nice tip, thanks! i'll use an approach like this.

Reply:
on that note, just confirming: the checks i'm writing here are just for virtual computed columns. but it seems that Steven's checks are for all computed columns. should i only be checking virtual ones here?

Reply:
Yeah, you only need to check virtual computed columns. I think Steven's checks were overly restrictive.